technical-docs/k8ssandra-upfront-validations.md

```diff
@@ -0,0 +1,527 @@
+[SKIPPED] RELEASE-NAME-cass-operator apps/v1/Deployment: Skipped because the deployment is not targeted by a HorizontalPodAutoscaler
+
+```
+
+
+## config-lint
+
+Stelligent config-lint is an open source command line tool to lint configuration files in a variety of formats: JSON, YAML, Terraform, and Kubernetes.
+
+### Features:
+
+* Enables developers with abiltiy to validate configuration files.
+* Provides custom validations and built-in rules to ensure configurations meet best practices.
+* Custom files can be used for other formats.
+* Built-in rules provided for Terraform
+
+
+#### Example Kubernetes
+
+```yaml
+version: 1
+description: Rules for Kubernetes spec files
+type: Kubernetes
+files:
+  - "*.yml"
+rules:
+
+  - id: ALLOW_KIND
+    severity: FAILURE
+    message: Allowed kinds
+    resource: "*"
+    assertions:
+      - key: kind
+        op: in
+        value: Pod,Policy,ServiceAccount,NetworkPolicy
+    tags:
+      - kind
+
+  - id: POD_CONTAINERS
+    severity: FAILURE
+    message: Pod must include containers
+    resource: Pod
+    assertions:
+      - key: spec.containers
+        op: present
+    tags:
+      - pod
+
+  - id: POD_SECURITY_CONTEXT
+    severity: FAILURE
+    message: Pod should set securityContent
+    resource: Pod
+    assertions:
+      - key: spec.securityContext.runAsNonRoot
+        op: eq
+        value: true
+      - key: spec.securityContext.readOnlyRootFilesystem
+        op: eq
+        value: true
+    tags:
+      - pod
+      - security
+
+  - id: DEFAULT_NAMESPACE
+    severity: FAILURE
+    message: Policy should not use default namespace
+    resource: Policy
+    assertions:
+      - key: spec.namespace
+        op: ne
+        value: default
+    tags:
+      - policy
+
+  - id: NETWORK
+    severity: FAILURE
+    message: Network policy should include from pods
+    resource: NetworkPolicy
+    assertions:
+      - key: spec.allowIncoming.from[].pods
+        op: present
+    tags:
+      - network
+  - id: DOCKER_REGISTRY
+    severity: FAILURE
+    message: Pods should pull from one of these docker registries
+    resource: Pod
+    assertions:
+     - every:
+         key: spec.containers
+         expressions:
+           - or:
+             - key: image
+               op: starts-with
+               value: <private docker registry url 1>
+             - key: image
+               op: starts-with
+               value: <private docker registry url 2>
+    tags:
+      - pod
+```
+Example for generic value assertions
+
+```yaml
+
+rules:
+…
+
+- id: REQ_COLOR
+    message: Missing or invalid color
+    severity: FAILURE
+    resource: ui-form
+    assertions:
+      - key: color
+        op: in
+        value: red,blue,green
+
+```
+
+
+## kubeval
+
+    TODO
+
+
+## Pros / Cons
+
+    TODO
+
+### config-lint
+
+Pros: Fills in for custom solutions and complements were kube-score leaves off.
+Cons: TODO
+
+### kube-score
+
+Pros: TODO
+Cons: TODO
```

Comment:
Does kube-score allow to create custom checks that would solve our heap settings issue? It doesn't seem so.

Reply:
No, that is why config-lint is being considered to be used along with kube-score.